src/tablestore.rs

```diff
@@ -403,17 +479,27 @@ impl TableStore {
         }
     }
 
-    fn parse_id(&self, path: &Path, expected_extension: &str) -> Result<u64, SlateDBError> {
-        match path.extension() {
-            Some(ext) if ext == expected_extension => path
-                .filename()
-                .expect("invalid wal file")
-                .split('.')
-                .next()
-                .ok_or_else(|| SlateDBError::InvalidDBState)?
-                .parse()
-                .map_err(|_| SlateDBError::InvalidDBState),
-            _ => Err(SlateDBError::InvalidDBState),
+    /// Parses the SsTableId from a given path
+    fn parse_id(root_path: &Path, path: &Path) -> Result<Option<SsTableId>, SlateDBError> {
+        if let Some(mut suffix_iter) = path.prefix_match(root_path) {
+            let next_part = suffix_iter.next().map(|s| s.as_ref().to_owned());
+            match next_part.as_deref() {
+                Some("wal") => suffix_iter
+                    .next()
+                    .and_then(|s| s.as_ref().split('.').next().map(|s| s.parse::<u64>()))
+                    .transpose()
+                    .map(|r| r.map(SsTableId::Wal))
+                    .map_err(|_| SlateDBError::InvalidDBState),
+                Some("compacted") => suffix_iter
+                    .next()
+                    .and_then(|s| s.as_ref().split('.').next().map(Ulid::from_string))
+                    .transpose()
+                    .map(|r| r.map(SsTableId::Compacted))
+                    .map_err(|_| SlateDBError::InvalidDBState),
+                _ => Ok(None),
+            }
```

Comment:
Consider:

```rust
            match suffix_iter.next() {
                Some(a) if a.as_ref() == "wal" => suffix_iter
                    .next()
                    .and_then(|s| s.as_ref().split('.').next().map(|s| s.parse::<u64>()))
                    .transpose()
                    .map(|r| r.map(SsTableId::Wal))
                    .map_err(|_| SlateDBError::InvalidDBState),
                Some(a) if a.as_ref() == "compacted" => suffix_iter
                    .next()
                    .and_then(|s| s.as_ref().split('.').next().map(Ulid::from_string))
                    .transpose()
                    .map(|r| r.map(SsTableId::Compacted))
                    .map_err(|_| SlateDBError::InvalidDBState),
                _ => Ok(None),
            }
```

It's still a little clunky (we are at the mercy of `object_store`'s API), but I find it more readable overall, and as a bonus it avoids allocating a `String` via `.as_ref().to_owned()`.